hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/Text.java

```diff
@@ -301,9 +305,18 @@ public void clear() {
    */
   private boolean ensureCapacity(final int capacity) {
     if (bytes.length < capacity) {
+      // use long to allow overflow
+      long tmpLength = bytes.length;
+      long tmpCapacity = capacity;
+
       // Try to expand the backing array by the factor of 1.5x
-      // (by taking the current size + diving it by half)
-      int targetSize = Math.max(capacity, bytes.length + (bytes.length >> 1));
+      // (by taking the current size + diving it by half).
+      //
+      // If the calculated value is beyond the size
+      // limit, we cap it to ARRAY_MAX_SIZE
+      int targetSize = (int)Math.min(ARRAY_MAX_SIZE,
```

Comment:
I would separate it into two lines.
int targetSize = Math.max(tmpCapacity, tmpLength + (tmpLength >> 1));
targetSize = (int)Math.min(ARRAY_MAX_SIZE, targetSize);

Reply:
Why do we need the cast to int by the way?

Reply:
The overloaded Math.max(long, long) will be used, which returns long, so it has to be cast back to an int. Therefore, in your suggestion, targetSize should also be a long (which, again has to be cast to an "int" when you instantiate the array).

Reply:
I would prefer to cast inside so it uses Math.max(int, int) and keeps everything as ints.
The splitting for the min and the max is more a matter of readability.

Reply:
@goiri could you comment on the latest changes?

Reply:
I think the following is easier to read and less casting required:
```
int targetSize = (int) bytes.length + (bytes.length >> 1);
targetSize = Math.max(capacity, targetSize);
targetSize = Math.min(ARRAY_MAX_SIZE, targetSize);
```

Reply:
The problem is that `bytes.length + (bytes.length >> 1);` might overflow and end up being negative which results in choosing `capacity` all the time instead of `ARRAY_MAX_SIZE`. So we either temporarily store this as `long` and cast back to `int` or we directly check if `targetSize` is negative after `int targetSize = (int) bytes.length + (bytes.length >> 1);`.

Reply:
With the explanation it makes sense to me.

Reply:
If we are protecting against overflows, then we should make it explicit:
```
long targetSizeLong = bytes.length + (bytes.length >> 1);
int targetSize = (int)Math.min(targetSizeLong, ARRAY_MAX_SIZE);
targetSize = Math.max(capacity, targetSize);
```